source/_drafts/Feature-Spotlight-Middleware.md

```diff
@@ -0,0 +1,185 @@
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
+For example, you have both ads and content in your HLS playlist but you want the controls to have a different times for each.
+When the content video is playing you want to display a timeline of 5 minutes. And for ads, you want to display 30 seconds for the pre-roll.
+Yet, currently, the timeline is showing a combined duration of 5 minutes and 30 seconds.
+So, you can add a middleware that knows when the ad is being played and tells the player that the duration is 30 seconds
+and when the content is playing that the duration is 5 minutes.
+```js
+// register a star-middleware because HLS has two mimetypes
+videojs.use('*', function(player) {
+  return {
+    setSource(srcObj, next) {
+      const type = srcObj.type;
+
+      if (type !== 'application/x-mpegurl' && type !== 'application/vnd.apple.mpegurl') {
+
+        // call next with an error to signal you cannot handle the source
+        next(new Error('Source is not an HLS source'));
+      } else {
+
+        // in here we know we're playing back an HLS source.
+        // We don't want to do anything special for it, so, pass along the source along with a null.
+        next(null, srcObj);
+      }
+    },
+
+    // this method gets called on the tech and then up the middleware chain providing the values as you go along
+    duration(durationFromTech) {
+      if (areWeCurrentlyPlayingAnAd(durationFromTech)) {
+        // since we're now in an ad, return the ad duration
+        return 30;
+      } else {
+        // we're playing back content, so, return that duration
+        return 5 * 60;
+      }
+    }
+  }
+});
+```
+
+A simple but interesting middleware to look at is the [playbackrate-adjuster][pra].
+The middleware will change the times of the controls depending on the current rate.
+If you're playing back a 20 minute video and change the rate to 2x, the controls will adjust to display 10 minutes.
+Let's take a look at the code.
+```js
+videojs.use('*', function(player) {
+  /* ... */
+
+  return {
+    setSource(srcObj, next) {
+      next(null, srcObj);
+    },
+
+    duration(dur) {
+      return dur / player.playbackRate();
+    },
+
+    /* ... */
+  };
+});
+```
+So, here, we attach a star-middleware because we want to have it applied to all our videos.
+In `setSource`, we also call `next` direct will `null` and the `srcObj` because we can handle any source.
+We also set up our `duration` method to take in the duration from the previous middleware and divide it by the playback rate we get from the player.
+
+If you look at the [code][] you can see some other methods next to duration.
+They're there to make sure other methods that rely on timing get updated.
+The two main ones to note is that there's a `currentTime` and a `setCurrentTime`.
```

Comment:
This sentence doesn't really read correctly. I think it's the plural `ones` and singular verb `is`. Maybe: 

> The two methods to notice are `currentTime` and `setCurrentTime`.